### PR TITLE
[api] Recovery api panic

### PR DESCRIPTION
**Pull request by envestcc**
# Description

as title

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by CoderZhi — APPROVED** [review]

File: api/web3server.go [hunk 1]
```diff
@@ -104,7 +105,7 @@ func NewWeb3Handler(core CoreService, cacheURL string, batchRequestLimit int) We
 }
 
 // HandlePOSTReq handles web3 request
-func (svr *web3Handler) HandlePOSTReq(ctx context.Context, reader io.Reader, writer apitypes.Web3ResponseWriter) error {
+func (svr *web3Handler) HandlePOSTReq(ctx context.Context, reader io.Reader, writer apitypes.Web3ResponseWriter) (err error) {
```
Comment (dustinxie) [inline api/web3server.go:108]: no need to add `err` since it'll be defined at L111?

Comment (envestcc) [reply api/web3server.go:108]: It's used for defer func, only in this way you can actually change the returned error in the defer function

File: api/web3server.go [hunk 2]
```diff
@@ -114,6 +115,15 @@ func (svr *web3Handler) HandlePOSTReq(ctx context.Context, reader io.Reader, wri
 		_, err = writer.Write(&web3Response{err: err})
 		return err
 	}
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Wrapf(errPanic, "recovered from panic: %v, request params: %+v", r, web3Reqs)
+			return
+		}
+		if err != nil {
+			err = errors.Wrapf(err, "failed to handle web3 requests: %+v", web3Reqs)
```
[the remark this answers is not in the document]
Comment (envestcc) [reply api/web3server.go:124]: I think so, it's the last statement in the defer closure

[the remark this answers is not in the document]
Comment (dustinxie) [reply api/web3server.go:124]: should modify all `return xxx` to below?
```
err = xxx
return
```

**Review by dustinxie — APPROVED** [review]

**Pull request merged**